### DISCUSSION TITLE
Matrix4x4.CreateTransformation(Vector3, Quaternion, Vector3)

### DISCUSSION THREAD

**Discussion by CaseyHofland** [dotnet/runtime#108822] [Ideas]
Creating a transformation matrix is not really an easy task, and it would be great if a standardized implementation could be provided by .Net.

Currently I'm using this:
```csharp
public static partial class Matrix
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Matrix4x4 CreateTransformation4x4(Vector3 position, Quaternion quaternion, Vector3 scales)
        {
            var translationMatrix = Matrix4x4.CreateTranslation(position);
            var rotationMatrix = Matrix4x4.CreateFromQuaternion(quaternion);
            var scaleMatrix = Matrix4x4.CreateScale(scales);

            // Combine the matrices: scale -> rotate -> translate
            return scaleMatrix * rotationMatrix * translationMatrix;
        }
}
```

This works fine, but then came the questions of:
- Does this work with non-uniform scaling?
- What happens with impossible to resolve quaternions?
- NaN values?
- What if there is a center point?
- How optimized is it to create 3 matrixes to create 1 matrix? What about 7 (I think) matrixes when you have a centerpoint?

### **But most important of all: is this guaranteed to provide me with the correct matrix?**

All of these questions I searched, and non of the answers were even close to human-readable for a layman like me. My concern is that the longer this takes, the wronger will be the implementations used by the industry ([even Unity is guilty of this](https://discussions.unity.com/t/float4x4-trs-does-tsr-instead/892229)).

So I implore you: save us .NET gods.

**Comment by colejohnson66** [comment]
You could try making a proper API proposal in the issues tab
